### PR TITLE
tx polling

### PR DESCRIPTION
**Pull request by soshochang**

### CODE REVIEW

**Comment by scriptfans** [comment]
hi, how to deal with paginate？loadFromWallet can receive an optional pageNum parameter, when clicking the 'loading more' button, we need to pass it. @soshochang

**Comment by scriptfans** [comment]
So I think implement pull-down refresh is the best solution.

File: src/popup/pages/AccountDetail/index.tsx [hunk 1]
```diff
@@ -17,14 +17,25 @@ interface IProps {
...
+  private static GET_TX_INTERVAL_MS: number = 10000;
+  private getTransactionsInterval?: number = undefined;
+
   public handleTabChange = (_: object, idx: number) => {
     this.props.store.accountDetailStore.activeTabIdx = idx;
   }
 
   public componentDidMount() {
     this.props.store.accountDetailStore.loadFromWallet();
+    this.getTransactionsInterval = setInterval(() => {
```
Comment (dwalintukan) [inline src/popup/pages/AccountDetail/index.tsx]: move this entire logic in didMount to the store and call init() here:
```
@action
public init = () => {
   loadFromWallet()
   setInterval()
}
```

File: src/popup/pages/AccountDetail/index.tsx [hunk 2]
```diff
@@ -17,14 +17,25 @@ interface IProps {
 
 @inject('store')
 @observer
-
 class AccountDetail extends Component<WithStyles & IProps, {}> {
+  private static GET_TX_INTERVAL_MS: number = 10000;
+  private getTransactionsInterval?: number = undefined;
```
Comment (dwalintukan) [inline src/popup/pages/AccountDetail/index.tsx]: move both vars to store. rule of thumb is to try to keep all biz logic out of components. components should be as "dumb" as possible, get data from store and display.

File: src/popup/pages/AccountDetail/index.tsx [hunk 3]
```diff
@@ -17,14 +17,25 @@ interface IProps {
...
 
   public componentDidMount() {
     this.props.store.accountDetailStore.loadFromWallet();
+    this.getTransactionsInterval = setInterval(() => {
+      this.props.store.accountDetailStore.loadFromWallet();
+    }, AccountDetail.GET_TX_INTERVAL_MS);
+  }
+
+  public componentWillUnmount() {
+    if (this.getTransactionsInterval) {
```
Comment (dwalintukan) [inline src/popup/pages/AccountDetail/index.tsx]: move to store

File: src/popup/pages/AccountDetail/index.tsx [hunk 4]
```diff
@@ -92,32 +103,32 @@ const shortenTxid = (txid?: string) => {
 
 const TransactionList = ({ classes, transactions }: any) =>
   transactions.map(({ id, pending, confirmations, timestamp, amount }: Transaction) => (
-    <ListItem divider key={id} className={classes.listItem}>
```
Comment (dwalintukan) [inline src/popup/pages/AccountDetail/index.tsx]: This indentation was correct.

File: src/popup/pages/AccountDetail/index.tsx [hunk 5]
```diff
@@ -92,32 +103,32 @@ const shortenTxid = (txid?: string) => {
...
+    <AmountInfo classes={classes} amount={amount} token="QTUM" />
+    <div>
+      <KeyboardArrowRight className={classes.arrowRight} />
+    </div>
+  </ListItem>
+));
 
 const TokenList = ({ classes, tokens }: any) =>
   tokens.map(({ name, token, amount, url }: any) => (
-    <ListItem divider key={token} className={classes.listItem} onClick={() => window.open(url, '_blank')}>
```
Comment (dwalintukan) [inline src/popup/pages/AccountDetail/index.tsx]: should indent after an encapsulating bracket `(` to show containment

**Comment by dwalintukan** [comment]
@scriptfans Don't like the idea that the user has to keep manually refreshing to get new data on the page. What about if we implement a call to refresh up to the amount of pages showing?

eg.
If we are on page 3 of the tx list, every minute, we will call `getTransactions(1) + getTransactions(2) + getTransactions(3)`

File: src/stores/AccountDetailStore.ts [hunk 6]
```diff
@@ -7,19 +7,23 @@ import AppStore from './AppStore';
...
 export default class AccountDetailStore {
+  private static GET_TX_INTERVAL_MS: number = 60000;
+
   @observable public activeTabIdx: number = 0;
   @observable.shallow public items: Transaction[] = [];
   @observable public pageNum: number = 0;
   @observable public pagesTotal?: number;
 
-  constructor(private app: AppStore) {}
+  private getTransactionsInterval?: NodeJS.Timer = undefined;
```
Comment (dwalintukan) [inline src/stores/AccountDetailStore.ts]: move this under public vars (even computed). I like to do it this way:
```
@observable publics
@computed publics
publics

@observable privates
@computed privates
privates
```

File: src/stores/AccountDetailStore.ts [hunk 7]
```diff
@@ -7,19 +7,23 @@ import AppStore from './AppStore';
...
 
   @computed public get hasMore(): boolean {
     return !!this.pagesTotal && (this.pagesTotal > this.pageNum + 1);
- }
+  }
+
+  constructor(private app: AppStore) {}
```
Comment (dwalintukan) [inline src/stores/AccountDetailStore.ts]: dont think we need to set the visibility on a param? also, you need to define a class var for `app`

File: src/stores/AccountDetailStore.ts [hunk 8]
```diff
@@ -28,6 +32,19 @@ export default class AccountDetailStore {
     this.items = this.items.concat(txs);
   }
 
+  public startTxPolling = () => {
+    this.loadFromWallet();
+    this.getTransactionsInterval = setInterval(() => {
+      this.loadFromWallet();
```
Comment (dwalintukan) [inline src/stores/AccountDetailStore.ts]: This is missing the whole logic of checking the block number and you are not calling the previous pages. It should follow this logic:
1. on interval tick, call RPC `getblockcount`
2. if new block num:
3. you need to call `getTransactions()` for each page you are on. eg. if you are on page 3, you need to call `getTransactions(1)` + `getTransactions(2)` + `getTransactions(3)` and then concatenate the data into one list then return

Comment (soshochang) [reply src/stores/AccountDetailStore.ts]: Not sure if I am misunderstanding what is missing,  but the block count for my transactions does update, and it does update for all the pages that have been loaded with loadMore.
 I'm using this to do it by passing in this.pageNum:
@action public async loadFromWallet() {
    this.items = await this.load(this.pageNum);
  }

File: src/stores/AccountDetailStore.ts (hunk 7 above)
Comment (soshochang) [reply src/stores/AccountDetailStore.ts]: as discussed, this is alternate syntax for class var declaration

File: src/stores/AccountDetailStore.ts [hunk 9]
```diff
@@ -7,33 +7,62 @@ import AppStore from './AppStore';
...
   }
 
+  public startTxPolling = () => {
+    this.loadFromWallet();
+    this.getTransactionsInterval = setInterval(() => {
+      this.refreshTransactions();
+    }, AccountDetailStore.GET_TX_INTERVAL_MS);
+  }
+
+  public stopTxPolling = () => {
```
Comment (dwalintukan) [inline src/stores/AccountDetailStore.ts]: Should be an action since it's modifying state

File: src/stores/AccountDetailStore.ts [hunk 10]
```diff
@@ -7,33 +7,62 @@ import AppStore from './AppStore';
...
+      this.pageNum = 0;
+    }
+  }
+
+  // TODO - if a new transaction comes in, the transactions on a page will shift(ie if 1 page has 10 transactions, transaction number 10 shifts to page2), and the bottom most transaction would disappear from the list. Need to add some additional logic to keep the bottom most transaction displaying.
+  @action
+  public async refreshTransactions() {
+    let refreshedItems: Transaction[] = [];
+    let txs: Transaction[];
+    for (let i = 0; i <= this.pageNum; i++) {
```
Comment (dwalintukan) [inline src/stores/AccountDetailStore.ts]: Is this a zero index call to getTransactions? Should page 0 be page 1?

File: src/stores/AccountDetailStore.ts [hunk 11]
```diff
@@ -7,33 +7,62 @@ import AppStore from './AppStore';
...
+      txs = await this.load(i);
+      refreshedItems = refreshedItems.concat(txs);
```
Comment (dwalintukan) [inline src/stores/AccountDetailStore.ts]: Inline the txs call so we don't need to store in the temp list. `.concat(await this.load(I))`

File: src/stores/AccountDetailStore.ts [hunk 12]
```diff
@@ -7,33 +7,62 @@ import AppStore from './AppStore';
...
+  constructor(private app: AppStore) {}
 
   @action public async loadFromWallet() {
     this.items = await this.load();
   }
 
   @action public async loadMore() {
     const txs = await this.load(this.pageNum + 1);
-
+    this.pageNum = this.pageNum + 1;
```
Comment (dwalintukan) [inline src/stores/AccountDetailStore.ts]: Move this above the load call and use this.pageNum so you don't have to increment in two places

File: src/stores/AccountDetailStore.ts (hunk 10 above)
Comment (soshochang) [reply src/stores/AccountDetailStore.ts]: yea it is a 0 index call, pageNum starts at 0 and that gives us the first page

**Pull request merged**